sympy/logic/tests/test_lra_theory.py

```diff
@@ -0,0 +1,297 @@
+    eps = sympify(1)
+
+    for _ in range(iter):
+        assign = {key: val[0] + val[1]*eps for key, val in assignment.items()}
+        try:
+            for cons in constr:
+                assert cons.subs(assign) == True
+            return assign
+        except AssertionError:
+            eps = eps/2
+
+    return None
+
+
+def test_from_encoded_cnf():
+    s1, s2 = symbols("s1 s2")
+
+    # Test preprocessing
+    # Example is from section 3 of paper.
+    phi = Q.prime(a) & (x >= 0) & ((x + y <= 2) | (x + 2 * y - z >= 6)) & (Eq(x + y, 2) | (x + 2 * y - z > 4))
+    cnf = CNF.from_prop(phi)
+    enc = EncodedCNF()
+    enc.from_cnf(cnf)
+    lra, x_subs, s_subs = LRASolver.from_encoded_cnf(enc)
+    assert lra.A.shape == (2, 5)
+    assert str(lra.slack) == '[_s1, _s2]'
+    assert str(lra.nonslack) == '[_x1, _x2, _x3]'
+    assert lra.A == Matrix([[ 1,  1, 0, -1,  0],
+                            [-1, -2, 1,  0, -1]])
+    assert {(str(b.var), b.bound, b.upper, b.equality, b.strict) for b in lra.boundry_enc.values()} == {('_s1', 2, None, True, False),
+    ('_s1', 2, True, False, False),
+    ('_s2', -4, True, False, True),
+    ('_s2', -6, True, False, False),
+    ('_x1', 0, False, False, False)}
+
+    # test functions
+    g = Function('g')(x)
+    f = Function('f')()
+    phi = (g + x + f + g**2 <= 2)
+    cnf = CNF.from_prop(phi)
+    enc = EncodedCNF()
+    enc.from_cnf(cnf)
+    lra, x_subs, s_subs = LRASolver.from_encoded_cnf(enc)
+    assert str(lra.slack) == '[_s1]'
+    # TODO: fix bug with constant functions so assert statement passes
+    #assert str(lra.nonslack) == '[_x1, _x2, _x3, _x4]'
+    #assert lra.A == Matrix([[1, 1, 1, 1, -1]])
+
+    # test no constraints
+    phi = Q.prime(x) & Q.integer(y)
+    cnf = CNF.from_prop(phi)
+    enc = EncodedCNF()
+    enc.from_cnf(cnf)
+    lra, x_subs, s_subs = LRASolver.from_encoded_cnf(enc)
+    assert lra.A == Matrix()
+    assert lra.slack == []
+    assert lra.nonslack == []
+
+
+def test_LRA_solver():
+    s1, s2 = symbols("s1 s2")
+    # Empty matrix should be handled.
+    # If the preprocessing step doesn't do anything, then the matrix is empty.
+    phi = (x >= 0) & (x >= 1) & (x <= -1)
+    cnf = CNF.from_prop(phi)
+    enc = EncodedCNF()
+    enc.from_cnf(cnf)
+    lra, x_subs, s_subs = LRASolver.from_encoded_cnf(enc)
+
+
+    assert len(lra.A) == 0
+    assert lra.assert_enc_boundry(enc.symbols.index((Q.ge(x, 0))) +1) == ("OK", None)
+    assert lra.assert_enc_boundry(enc.symbols.index((Q.ge(x, 1))) +1) == ("OK", None)
+    assert lra.assert_enc_boundry(enc.symbols.index((Q.le(x, -1))) + 1) == ('UNSAT', {-(enc.symbols.index((Q.le(x, -1))) + 1), -(enc.symbols.index((Q.ge(x, 1))) +1)})
+    # assert lra.assert_con(Q.ge(x, 0)) == ("OK", None)
+    # assert lra.assert_con(Q.ge(x, -1)) == ('SAT', {x: 0, y: 0})
+    # assert lra.assert_con(Q.le(x, -1)) == ('UNSAT', {x <= -1, x >= 0})
+    #
+    # m = Matrix()
+    # lra = LRASolver(m, [], [x, y])
```

Comment:
This commented out code code uses depreciated functions. I'm not sure if I should convert it into code that uses the new functions or just get rid of it. I feel like it's kind of pointless because test_random_problems() should be able to handle all of the cases here.